README.md
# gender_selection

A Flutter package for gender selection. It is an aweome gender selection widget with cool gradients and animation effects

[![pub package](https://img.shields.io/badge/pub-0.5.0-orange.svg)](https://pub.dartlang.org/packages/gender_selection)
[![Awesome Flutter](https://img.shields.io/badge/Awesome-Flutter-blue.svg?longCache=true&style=flat-square)](https://github.com/404shades)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/404shades)


## Installing
Add this to your package's pubspec.yaml file:
```dart
dependencies:
gender_selection: "^0.0.7"
```

## Simple Usage

```dart
import 'package:gender_selection/gender_selection.dart';

GenderSelection(
maleText: "", //default Male
femaleText: "", //default Female
linearGradient: pinkRedGradient,
selectedGenderIconBackgroundColor: Colors.indigo, // default red 
checkIconAlignment: Alignment.centerRight,   // default bottomRight
selectedGenderCheckIcon: null, // default Icons.check
onChanged: (Gender gender){
print(gender);
},
equallyAligned: true,
animationDuration: Duration(milliseconds: 400),
isCircular: true, // default : true,
isSelectedGenderIconCircular: true,
opacityOfGradient: 0.6,
padding: const EdgeInsets.all(3),
size: 120, //default : 120

),
```
## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.0.1] - Release 07 November 2019

* A Flutter package for Gender Selection written purely in darts. Works in iOS and Android both

## [0.0.2]- Release 07 November 2019

* Updated Asset Image to load male and female default images

## [0.0.3]- Release 07 November 2019

* Adjust the selected icon alignment accordingly

## [0.0.4]- Release 07 November 2019

* Fix issue for no onChanged given

## [0.0.5]- Release 07 November 2019

* Now you can remove icon by passing selectedGenderCheckIcon as null

## [0.0.6]- Release 07 November 2019

* Fix bug for null icon data

## [0.0.7]- Release 07 November 2019

* Fix bug for null icon data

## [0.0.8]- Release 07 November 2019

* Updated with example and readme


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:gender_selection/gender_selection.dart';
import 'package:example/utils/gradients.dart';
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Gender Selection Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        primaryColor: Colors.red,
      ),
      home: HomeScreen()
    );
  }
}

class HomeScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Gender Selection Demo"),
      ),
      body: SingleChildScrollView(
        child: Container(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              GenderSelection(
                femaleImage: NetworkImage("https://cdn1.iconfinder.com/data/icons/website-internet/48/website_-_female_user-512.png"),
                maleImage: NetworkImage("https://icon-library.net/images/avatar-icon/avatar-icon-4.jpg"),
                selectedGenderIconBackgroundColor: Colors.amber,
                selectedGenderTextStyle: TextStyle(color: Colors.amber,fontSize: 19,fontWeight: FontWeight.bold),
              ),
              GenderSelection(
                maleText: "",
                femaleText: "",
                linearGradient: pinkRedGradient,
                selectedGenderIconBackgroundColor: Colors.indigo,
                checkIconAlignment: Alignment.centerRight,
                selectedGenderCheckIcon: null,
                onChanged: (Gender gender){
                  print(gender);
                },
                equallyAligned: true,
                animationDuration: Duration(milliseconds: 400),
                isCircular: true,
                isSelectedGenderIconCircular: true,
                opacityOfGradient: 0.6,
                padding: const EdgeInsets.all(3),
                size: 120,
                
              ),
              GenderSelection(
                maleImage: AssetImage("assets/images/male.png"),
                femaleImage: AssetImage("assets/images/female.png"),
                selectedGenderIconBackgroundColor: Colors.green,
                selectedGenderIconColor: Colors.white,
                linearGradient: greenSexyGradient,
                onChanged: (Gender gender){
                  print(gender);
                },

              ),
              GenderSelection(
                isCircular: false,
                selectedGenderIconBackgroundColor: Colors.indigo,
                linearGradient: skyBlueGradient,
                onChanged: (Gender gender){
                  print(gender);
                },
              ),
            ],
          ),
        ),
      )
    );
  }
}






example/lib/utils/gradients.dart
import 'package:flutter/material.dart';


var bluePinkGradient = new LinearGradient(
  colors: [const Color(0xFFCC0099),const Color(0xFF6600FF)],
  tileMode: TileMode.clamp,
  begin: Alignment.topLeft,
  end:Alignment.bottomRight,
  stops: [0.0,1.0]
);

var blueGreenGradient = new LinearGradient(
  colors: [const Color(0xFF0099CC),const Color(0xFF99FF66)],
  tileMode: TileMode.clamp,
  begin: Alignment.topLeft,
  end: Alignment.bottomRight,
  stops: [0.0,1.0],

);
var yellowOrangeGradient = new LinearGradient(
  colors: [const Color(0xFFFFFF00),const Color(0xFFFF6600)],
  tileMode: TileMode.clamp,
  begin: Alignment.topLeft,
  end:Alignment.bottomRight,
  stops: [0.0,1.0],
);
var pinkRedGradient = new LinearGradient(
  colors: [const Color(0xFFFF3399),const Color(0xFFFF0000)],
  tileMode: TileMode.clamp,
  begin: Alignment.topLeft,
  end:Alignment.bottomRight,
  stops:[0.0,1.0]
);

var blackBlueGradient = new LinearGradient(
  colors: [Colors.blueGrey.shade800,Colors.black87],
  tileMode: TileMode.clamp,
  begin: Alignment.topLeft,
  end: Alignment.bottomRight,
  stops: [0.0,1.0]
);
var redSexyGradient = new LinearGradient(
  colors: [const Color(0xFFF6356F),const Color(0xFFFF5F50)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var greenSexyGradient = new LinearGradient(
  colors: [const Color(0xFF92D000),const Color(0xFFE1EB01)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var violetSexyGradient = new LinearGradient(
  colors: [const Color(0xFF52A7EA),const Color(0xFF712E98)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var blueSexyGradient = new LinearGradient(
  colors: [const Color(0xFF46A3B7),const Color(0xFF86F1DE)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var orangeSexyGradient = new LinearGradient(
  colors: [const Color(0xFFFF9B30),Colors.deepOrange],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var thodaSexyGradient = new LinearGradient(
  colors: [const Color(0xFFF58573),const Color(0xFF5F79F4)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var skyBlueGradient = new LinearGradient(
  colors: [const Color(0xFF15EDED),const Color(0xFF029CF5)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var blackSexyGradient = new LinearGradient(
  colors: [const Color(0xFF90203F),const Color(0xFF537895)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);

var disabledSilverGradient = new LinearGradient(
  colors: [const Color(0xFFF6F2F2),const Color(0xFFD9D9D9)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var balbirGradient = new LinearGradient(
  colors: [const Color(0xFFFF5F6d),const Color(0xFFFFC371)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);

var orangeCoral = new LinearGradient(
  colors: [const Color(0xFFff9966),const Color(0xFFff5e62)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);
var whiteGradient = new LinearGradient(
  colors: [const Color(0xFFfdfbfb),const Color(0xFFebedee)],
  tileMode: TileMode.clamp,
  begin: Alignment.bottomLeft,
  end: Alignment.topRight,
  stops: [0.0,1.0]
);

lib/gender_selection.dart
library gender_selection;

import 'package:flutter/material.dart';
import 'package:gender_selection/utils/gradients.dart';

enum Gender { Male, Female }

// ignore: must_be_immutable
class GenderSelection extends StatefulWidget {
  Gender selectedGender;
  final LinearGradient linearGradient;
  final ImageProvider maleImage;
  final ImageProvider femaleImage;
  final ValueChanged<Gender> onChanged;
  final String maleText;
  final bool equallyAligned;
  final String femaleText;
  final IconData selectedGenderCheckIcon;
  final double size;
  final bool isCircular;
  final EdgeInsetsGeometry padding;
  final Duration animationDuration;
  final double opacityOfGradient;
  final double selectedGenderIconSize;
  final bool isSelectedGenderIconCircular;
  final Color selectedGenderIconColor;
  final Color selectedGenderIconBackgroundColor;
  final TextStyle selectedGenderTextStyle;
  final TextStyle unSelectedGenderTextStyle;
  final Alignment checkIconAlignment;

  GenderSelection(
      {Key key,
      this.selectedGenderTextStyle = const TextStyle(
        fontSize: 19,
        color: Colors.redAccent,
        fontWeight: FontWeight.w600,
      ),
      this.unSelectedGenderTextStyle = const TextStyle(
        fontSize: 19,
        color: Colors.black,
        fontWeight: FontWeight.w600,
      ),
      this.selectedGenderIconBackgroundColor = Colors.red,
      this.selectedGenderIconColor = Colors.white,
      this.selectedGenderIconSize = 20.0,
      this.isSelectedGenderIconCircular = true,
      this.animationDuration = const Duration(milliseconds: 500),
      this.opacityOfGradient = 0.6,
      this.padding = const EdgeInsets.all(10),
      this.isCircular = true,
      this.size = 120.0,
      this.selectedGenderCheckIcon = Icons.check,
      this.equallyAligned = true,
      this.selectedGender,
      this.checkIconAlignment = Alignment.bottomCenter,
      @required this.onChanged,
      this.linearGradient = balbirGradient,
      this.maleImage = const AssetImage("assets/images/boy.png",
          package: 'gender_selection'),
      this.femaleImage = const AssetImage("assets/images/girl.png",
          package: 'gender_selection'),
      this.maleText = "Male",
      this.femaleText = "Female"})
      : super(key: key);
  @override
  _GenderSelectionState createState() => _GenderSelectionState();
}

class _GenderSelectionState extends State<GenderSelection> {
  @override
  Widget build(BuildContext context) {
    return FittedBox(
      fit: BoxFit.fitWidth,
      child: Padding(
        padding: widget.padding,
        child: Container(
          width: MediaQuery.of(context).size.width,
          child: Row(
            mainAxisSize: MainAxisSize.max,
            mainAxisAlignment: widget.equallyAligned
                ? MainAxisAlignment.spaceEvenly
                : MainAxisAlignment.center,
            children: <Widget>[
              GestureDetector(
                onTap: () {
                  setState(() {
                    widget.selectedGender = Gender.Male;
                    if (widget.onChanged != null) {
                      widget.onChanged(widget.selectedGender);
                    } else {
                      return;
                    }
                  });
                },
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Stack(
                      fit: StackFit.loose,
                      alignment: widget.checkIconAlignment,
                      children: <Widget>[
                        Container(
                          height: widget.size,
                          width: widget.size,
                          margin: const EdgeInsets.only(top: 20, bottom: 10),
                          decoration: BoxDecoration(
                            shape: widget.isCircular
                                ? BoxShape.circle
                                : BoxShape.rectangle,
                            image: DecorationImage(
                              image: widget.maleImage,
                              fit: BoxFit.cover,
                            ),
                          ),
                          child: AnimatedOpacity(
                            opacity: widget.selectedGender == Gender.Male
                                ? widget.opacityOfGradient
                                : 0.0,
                            duration: widget.animationDuration,
                            child: widget.selectedGender == Gender.Male
                                ? Container(
                                    decoration: BoxDecoration(
                                        gradient: widget.linearGradient,
                                        shape: widget.isCircular
                                            ? BoxShape.circle
                                            : BoxShape.rectangle),
                                  )
                                : null,
                          ),
                        ),
                        Positioned(
                          child: widget.selectedGenderCheckIcon != null
                              ? AnimatedContainer(
                                  duration: widget.animationDuration,
                                  height: widget.selectedGender == Gender.Male
                                      ? widget.selectedGenderIconSize
                                      : 0.0,
                                  width: widget.selectedGender == Gender.Male
                                      ? widget.selectedGenderIconSize
                                      : 0.0,
                                  padding: const EdgeInsets.all(1),
                                  curve: Curves.fastLinearToSlowEaseIn,
                                  decoration: BoxDecoration(
                                      shape: widget.isSelectedGenderIconCircular
                                          ? BoxShape.circle
                                          : BoxShape.rectangle,
                                      color: widget
                                          .selectedGenderIconBackgroundColor),
                                  child: widget.selectedGender == Gender.Male
                                      ? LayoutBuilder(
                                          builder: (context, constraints) {
                                            return Icon(
                                              widget.selectedGenderCheckIcon,
                                              color: widget
                                                  .selectedGenderIconColor,
                                              size: constraints.maxHeight,
                                            );
                                          },
                                        )
                                      : null)
                              : Container(
                                  height: 0,
                                  width: 0,
                                ),
                        )
                      ],
                    ),
                    Text(widget.maleText,
                        style: widget.selectedGender == Gender.Male
                            ? widget.selectedGenderTextStyle
                            : widget.unSelectedGenderTextStyle)
                  ],
                ),
              ),
              GestureDetector(
                onTap: () {
                  setState(() {
                    widget.selectedGender = Gender.Female;
                    if (widget.onChanged != null) {
                      widget.onChanged(widget.selectedGender);
                    } else {
                      return;
                    }
                  });
                },
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Stack(
                      fit: StackFit.loose,
                      alignment: widget.checkIconAlignment,
                      children: <Widget>[
                        Container(
                          height: widget.size,
                          width: widget.size,
                          margin: const EdgeInsets.only(top: 20, bottom: 10),
                          decoration: BoxDecoration(
                              shape: widget.isCircular
                                  ? BoxShape.circle
                                  : BoxShape.rectangle,
                              image: DecorationImage(
                                  image: widget.femaleImage,
                                  alignment: Alignment.center,
                                  fit: BoxFit.cover)),
                          child: AnimatedOpacity(
                            opacity: widget.selectedGender == Gender.Female
                                ? widget.opacityOfGradient
                                : 0.0,
                            duration: widget.animationDuration,
                            child: widget.selectedGender == Gender.Female
                                ? Container(
                                    decoration: BoxDecoration(
                                        gradient: widget.linearGradient,
                                        shape: widget.isCircular
                                            ? BoxShape.circle
                                            : BoxShape.rectangle),
                                  )
                                : null,
                          ),
                        ),
                        Positioned(
                          child: widget.selectedGenderCheckIcon != null
                              ? AnimatedContainer(
                                  duration: widget.animationDuration,
                                  height: widget.selectedGender == Gender.Female
                                      ? widget.selectedGenderIconSize
                                      : 0.0,
                                  width: widget.selectedGender == Gender.Female
                                      ? widget.selectedGenderIconSize
                                      : 0.0,
                                  padding: const EdgeInsets.all(1),
                                  curve: Curves.fastLinearToSlowEaseIn,
                                  decoration: BoxDecoration(
                                      shape: widget.isSelectedGenderIconCircular
                                          ? BoxShape.circle
                                          : BoxShape.rectangle,
                                      color: widget
                                          .selectedGenderIconBackgroundColor),
                                  child: widget.selectedGender == Gender.Female
                                      ? LayoutBuilder(
                                          builder: (context, constraints) {
                                            return Icon(
                                              widget.selectedGenderCheckIcon,
                                              color: widget
                                                  .selectedGenderIconColor,
                                              size: constraints.maxHeight,
                                            );
                                          },
                                        )
                                      : null)
                              : Container(
                                  height: 0,
                                  width: 0,
                                ),
                        )
                      ],
                    ),
                    Text(widget.femaleText,
                        style: widget.selectedGender == Gender.Female
                            ? widget.selectedGenderTextStyle
                            : widget.unSelectedGenderTextStyle)
                  ],
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}
